components/movementsensor/gpsutils/ntrip.go

```diff
@@ -256,13 +165,84 @@ func (n *NtripInfo) Connect(ctx context.Context, logger logging.Logger) error {
 	return err
 }
 
-// HasStream checks if the sourcetable contains the given mountpoint in it's stream.
-func (st *Sourcetable) HasStream(mountpoint string) (Stream, bool) {
-	for _, str := range st.Streams {
-		if str.MP == mountpoint {
-			return str, true
+// waitUntilCasterIsLive returns when n.client.IsCasterAlive() is true, or after 5 failed attempts.
+// This is a helper function intended only to be called from within Connect().
+func (n *NtripInfo) waitUntilCasterIsLive(logger logging.Logger) error {
+	if n.client.IsCasterAlive() {
+		return nil
+	}
+
+	logger.Infof("caster %s seems to be down, retrying", n.url)
+	// We will try to connect to the caster five times if it's down.
+	for attempts := 0; attempts < 5; attempts++ {
+		if n.client.IsCasterAlive() {
+			return nil
+		}
+		logger.Debugf("attempt(s) to connect to caster: %v ", attempts)
+	}
+	return fmt.Errorf("caster %s is down", n.url)
+}
+
+// GetStreamFromMountPoint attempts to connect to the NTRIP stream, store it in n.stream, and
+// return it. We give up and return an error after n.maxConnectAttempts unsuccessful tries.
+func (n *NtripInfo) GetStreamFromMountPoint(
+	cancelCtx context.Context,
+	logger logging.Logger,
+) (io.ReadCloser, error) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	success := false
+	attempts := 0
+
+	// Set the Timeout to 0 on the http client to prevent the ntrip stream from canceling itself.
+	// ntrip.NewClient() defaults sets this value to 15 seconds, which causes us to disconnect
+	// the ntrip stream and require a reconnection. Additionally, this should be tested with other
+	// CORS.
+	n.client.Timeout = 0
+
+	var rc io.ReadCloser
+	var err error
+
+	logger.Debug("Getting NTRIP stream")
+
+	for !success && attempts < n.maxConnectAttempts {
+		select {
+		case <-cancelCtx.Done():
+			return nil, errors.New("Canceled")
+		default:
+		}
+
+		rc, err = n.client.GetStream(n.MountPoint)
+		if err == nil {
+			success = true
 		}
+		attempts++
+	}
+
+	if err != nil {
+		logger.Errorf("Can't connect to NTRIP stream: %s", err)
+		return nil, err
 	}
+	logger.Debug("Connected to stream")
 
-	return Stream{}, false
+	n.stream = rc
+	return rc, nil
+}
+
+// Close shuts down all connections to the NTRIP caster.
+func (n *NtripInfo) Close(ctx context.Context) error {
```

Comment:
do we need this context? ntrip info doesn't have an interface its trying to match iirc

Reply:
We don't use it, but nearly every `Close()` in the codebase takes a context and returns an error, and I wanted to keep the interface consistent. but you're right that we don't need to match the component interface, and could change it. 

I don't feel strongly, and am happy to change it if you want.

Reply:
also don't feel strongly was just curious for the rationale